GENES = ["CMY",\
			"CTX-M",\
			"GES",\
			"IMP",\
			"KPC",\
			"NDM",\
			"PER",\
			"SHV",\
			"TEM",\
			"VEB",\
			"VIM"]
DB = ["NCBI", "CARD"]

FOCAL_GENES = ["CMY-2"]#,"GES-24","CTX-M-15","CTX-M-65","GES-24","IMP-1","KPC-2","NDM-1","PER-1","SHV-134","TEM-1","VIM-1"]

FOCAL_GENE_DICT = {"CMY-2": "CMY"}

# directory with fasta files containing genes of interest
FASTA_DIR = "/Users/Liam/Downloads/betalactamase-contigs"


rule all:
	input:
		expand("DB/plots/{db}-variants/{gene}.pdf", gene=GENES, db=DB),
		expand("DB/dists/{gene}_{db}.tsv", gene=GENES, db=DB),
		expand("analysis/{gene}_accessions.txt", gene=GENES),
		expand("analysis/sequence_assignments/{focal_gene}_{db}.csv", focal_gene=FOCAL_GENES, db=DB),
		expand("data/{db}_db.fa", db=DB),	
		expand("contigs/{gene}_combined_contigs.fa", gene=GENES),
		expand("analysis/sequence/{focal_gene}_seqs_extracted_from_contigs.fa", focal_gene=FOCAL_GENES),
		#expand("contigs/{focal_gene}/{focal_gene}_extracted.fa", focal_gene=FOCAL_GENES)
		expand("contigs/{focal_gene}/{focal_gene}_pangraph.json", focal_gene=FOCAL_GENES)

		#expand("DB/variant_fasta/{focal_gene}_{db}.fa", focal_gene=FOCAL_GENES, db=DB)

#############################
### SETTING UP DATABASES ####
#############################

rule download_NCBI_DB:
	output:
		"data/NCBI_db.fa"
	shell:
		"wget https://ftp.ncbi.nlm.nih.gov/pathogen/Antimicrobial_resistance/AMRFinderPlus/database/3.11/2022-12-19.1/AMR_CDS -O {output}"

rule download_CARD_DB:
	output:
		"data/CARD_db.fa"
	run:
		shell("wget https://card.mcmaster.ca/download/0/broadstreet-v3.2.6.tar.bz2")
		shell("tar -xvf broadstreet-v3.2.6.tar.bz2 ./nucleotide_fasta_protein_homolog_model.fasta")
		shell("rm broadstreet-v3.2.6.tar.bz2")
		shell("mv nucleotide_fasta_protein_homolog_model.fasta {output}")

rule extract_genes_DB:
	input:
		fasta="data/{db}_db.fa"
	params:
		gene_name="{gene}",
		db="{db}"
	output:
		"DB/gene_fasta/{gene}_{db}.fa"
	shell:
		"python scripts/extract_gene_DB.py {input.fasta} {params.gene_name} {params.db} {output}"

rule align_genes_DB:
	input:
		"DB/gene_fasta/{gene}_{db}.fa"
	output:
		"DB/gene_alns/{gene}_{db}.aln"
	shell:
		"mafft --quiet --auto {input} > {output}"

rule snp_dists_DB:
	input:
		"DB/gene_alns/{gene}_{db}.aln"
	output:
		"DB/gene_dists/{gene}_{db}.tsv"
	shell:
		"snp-dists -a -b {input} > {output}"

rule run_fasttree_DB:
	input:
		"DB/gene_alns/{gene}_{db}.aln"
	output:
		"DB/gene_trees/{gene}_{db}.nwk"
	shell:
		"FastTree -quiet -nt -gtr {input} > {output}"

rule refine_fasttree_DB:
	input:
		tree= "DB/gene_trees/{gene}_{db}.nwk",
		aln= "DB/gene_alns/{gene}_{db}.aln"
	output:
		"DB/refined-trees/{gene}_{db}.nwk"
	shell:
		"python scripts/refine_tree.py --tree_in {input.tree} --aln {input.aln} --tree_out {output}"

rule plot_tree_DB:
	input:
		"DB/genes_refined-trees/{gene}_{db}.nwk"
	output:
		"DB/genes_plots/{db}-variants/{gene}.pdf"
	shell:
		"Rscript scripts/plot_tree.R {input} {output}"

rule run_metadata:
	input:
		metadata="data/metadata.csv"
	params:
		gene="{gene}"
	output:
		"analysis/{gene}_accessions.txt"
	script:
		"scripts/metadata_gene_combinations.py"


rule copy_variant_fasta: # highly inelegant - copies files so is wasteful. But I don't understand symbolic links in snakemake
    input:
        lambda wildcards: f"DB/gene_fasta/{FOCAL_GENE_DICT[wildcards.focal_gene]}"+"_{db}.fa"
    output:
        "DB/variant_fasta/{focal_gene}_{db}.fa"
    shell:
        "cp {input} {output}"

# rule combine_fasta_files:
# 	input:

# rule split_focal_genes:
# 	input:
# 		"data/focal_genes.fasta"
# 	# params:
# 	# 	gene_name="{focal_gene}"
# 	# output:
# 	# 	"data/focal_genes/{focal_gene}.fa"
# 	shell:
# 		"python scripts/split_fasta.py --fasta {input} --outdir data/focal_genes/"


#############################
# GETTING CONTIG DATA READY #
#############################
rule combine_fastas_containing_gene:
	input:
		"analysis/{gene}_accessions.txt"
	output:
		"contigs/{gene}_combined_contigs.fa"
	run:
		shell("while read f; do cat "f"{FASTA_DIR}""/$f.fa >> {output}; done < {input}")


# rule extract_genes_from_contigs:
# 	input:
# 		gene_fasta="data/focal_genes/{focal_gene}.fa",
# 		input_fasta="test_{focal_gene}.fa"
# 	output:
# 		"analysis/sequence/{focal_gene}.fa"
# 	shell:
# 		"python scripts/extract_region_around_gene.py --gene {input.gene_fasta} --input_fasta {input.input_fasta} --output_fasta {output} --upstream 0 --downstream 0"
rule extract_genes_from_contigs:
	input:
		gene_fasta="data/focal_genes/{focal_gene}.fa",
		input_fasta= lambda wildcards:f"contigs/{FOCAL_GENE_DICT[wildcards.focal_gene]}"+"_combined_contigs.fa"
	output:
		"analysis/sequence/{focal_gene}_seqs_extracted_from_contigs.fa"
	shell:
		"python scripts/extract_region_around_gene.py --gene {input.gene_fasta} --input_fasta {input.input_fasta} --output_fasta {output} --upstream 0 --downstream 0"

rule name_observed_genes:
	input:
		fasta="analysis/sequence/{focal_gene}_seqs_extracted_from_contigs.fa",
		variants="DB/variant_fasta/{focal_gene}_{db}.fa"
	output:
		"analysis/sequence_assignments/{focal_gene}_{db}.csv"
	shell:
		"python scripts/name_variants.py --variant_fasta {input.variants} --output_file {output} --input_fasta {input.fasta}"

#############################
# RUNNING PANGRAPH PIPELINE #
#############################

rule extract_region_around_focal_gene:
	input:
		input_fasta=lambda wildcards:f"contigs/{FOCAL_GENE_DICT[wildcards.focal_gene]}"+"_combined_contigs.fa"
	params:
		focal_gene="data/focal_genes/{focal_gene}.fa",
		prefix="contigs/{focal_gene}/{focal_gene}_extracted"
	output:
		"contigs/{focal_gene}/{focal_gene}_extracted.fa"#,
		#"contigs/{focal_gene}/{focal_gene}_extracted.focal_gene.fa"
	shell: # to do: replace with extract_region_around_gene.py (I refactored slightly)
		"python scripts/extract_region_around_gene.py --gene {params.focal_gene} --input {input.input_fasta} --upstream 5000 --downstream 5000 --complete --output_fasta {params.prefix}.fa"



rule build_pangraph:
	input:
		"contigs/{focal_gene}/{focal_gene}_extracted.fa"
	output:
		"contigs/{focal_gene}/{focal_gene}_pangraph.json"
	shell:
		"pangraph build -k minimap2 --len 100 {input} > {output}"






