# Details of scripts

Unless otherwise specified, all should be run from within `scripts`.

##Â CARD data processing

`01_summarise-CARD-data.R`
_Status_: runs
_Usage_: `Rscript 01_summarise-CARD-data.R`
Needs: `index-for-model-sequences.txt` (CARD download, in `../CARD`)
      `../beta-lactamases.txt` (list of clinically relevant beta-lactamase families)
Makes: `gene-prevalence-ncbi_plasmid.csv` etc.
    `clinical-betalactamase-groups.pdf` and `.png` figure of prevalence of given families
Currently this makes the figure only for `ncbi_wgs`. It could be changed to do it for `ncbi_chromosome` and `ncbi_plasmid` as well (trivial).

`02_make-matrix-card-data.py`
_Status_: runs
_Usage_: `python 02_make-matrix-card-data.py`.
Needs: `card-genomes.txt` (CARD download, in `../CARD`)
Makes: `CARD-all-hits-presence-absence-chrom-plasmid.csv`
Could be improved to score perfect/strict matches differently. Currently both are stored as 1.

`03a_align-families.sh`
_Status_: runs
_Usage_: `source 03a_align-families.sh` (because of conda environments, `./03a_` won't work)
Needs: `nucleotide_fasta_protein_homolog_model.fasta` (CARD download, in `../CARD`)
Extracts gene sequences from this file with `grep` (could be replaced with python to be more reliable)
Then aligns these sequences with clustalo
Then uses `snp-dists -a -b` to get differences including insertions and deletions (gaps, not just SNPs)
(Could be put into python with calls, so that it can be merged with script below - but not necessary as it works now)

`03b_make-heatmap.py`
_Status_: runs
_Usage_: `python 03b_make-heatmap.py` (takes a few minutes; OXA is slow because large family)
Needs: `gene-prevalence-ncbi_wgs.csv` (or `ncbi_chromosome`, `ncbi_plasmid`). Those files are generated by `summarise-CARD-data.R`
`_diffs_aln.tsv` generated by `03a_align-families.sh`
Could save the clusters for a given distance threshold to file and/or compute the prevalence of each cluster. Might be interesting to add later.

## Downloading sequences

`download-gene-cluster.py`
_Status_: not tested
Needs:
`CARD-all-hits-presence-absence-chrom-plasmid.csv` (made by `make-matrix-card-data.py`, see below)
`card-genomes.txt` (CARD download, in `../CARD`)
`_diffs_aln.tsv` (from `make-heatmap.sh`)
Could be altered to take arguments with argparse.
Currently does not download the genomes.
Has getNeighbours within it.
Could call `ncbi-acc-download` from within script

## Processing

Once the fastas are all downloaded, this is what the pipeline looks like at the moment: something like

```bash
cat *.fa > all.fa
# Need to adapt extract-region.py to handle contigs which have more than one copy, and to have the option to store these contigs elsewhere
python ~/Dropbox/_Projects/pangraph/pangraph-tutorials/scripts/extract-region.py --gene ${gene}.fasta --input all.fa --upstream 5000 --downstream 5000 --complete --output all_u5k_d5k.fa
# unsure if --circular is appropriate here - depends on whether plasmids and chromosomes can be trusted to be circular  

# Pangraph
pangraph build all_u5k_d5k.fa > pangraph_all_u5k_d5k.json
pangraph export pangraph_all_u5k_d5k.json -p pangraph_all_u5k_d5k  -o ./

# Prepare gfa
python ~/Dropbox/_Projects/pangraph/pangraph-tutorials/scripts/prepare-pangraph-gfa.py pangraph_all_u5k_d5k.gfa

# Find the gene block in pangraph  
makeblastdb -in pangraph_all_u5k_d5k.fa -dbtype 'nucl'
blastn -query ${gene}.fasta -db pangraph_all_u5k_d5k.fa -outfmt 6

# Get the gene sequences
makeblastdb -in all_u5k_d5k.fa -dbtype 'nucl'
blastn -query ${gene}.fasta -db all.fa -outfmt "6 sseqid sseq" | sed 's/^/>/g' | sed -e 's/\t/\n/g' > gene_sequences.fa                                              

#


```

`extract-region.py`
Within `pangraph-tutorials` (maybe should rename directory - how does this affect links?)

`prepare-pangraph-gfa.py`
Within `pangraph-tutorials`

`compute-distances.py`
_Status_: not adapted yet from `CTX-M-55/true` example
Could compute SNP distances within python rather than snp-dists. 



### Visualization

`plot-breakpoint-dists-SNPs.R`
_Status_: not adapted yet from `CTX-M-55/true` example

## CARD data files

These are now in `beta-lactamases/CARD` since we don't need all CARD download files.

From CARD-prevalence download v3.1.0 [here](https://card.mcmaster.ca/download/6/prevalence-v3.1.0.tar.bz2):
`index-for-model-sequences.txt.gz`: contains all the detection statistics for the
sequences available in the above FASTA files, indicating pathogen, detection criteria,
ARO categorization, and similarity to curated CARD reference sequence

`card-genomes.txt.gz`: tab-separated; indicates for each analyzed genome which putative AMR
determinants were detected by RGI at both the Perfect and Strict confidence levels. This
provides a "putative resistome"  for each genome. Available and viewable on the web at:
 https://card.mcmaster.ca/genomes

From CARD-data download v3.2.3 [here](https://card.mcmaster.ca/download/0/broadstreet-v3.2.3.tar.bz2):

`nucleotide_fasta_protein_homolog_model.fasta`: Nucleotide and corresponding protein FASTA downloads are available as separate files for each model type.  For example, the "protein homolog" model type contains sequences of antimicrobial resistance genes that do not include mutation as a determinant of resistance - these data are appropriate for BLAST analysis of metagenomic data or searches excluding secondary screening for resistance mutations.
