# Details of scripts

Unless otherwise specified, all should be run from within `scripts`.
Remember to activate the conda environment first with `conda activate betalactamases`.

## CARD data processing

`01_summarise-CARD-data.R`
_Status_: runs
_Usage_: `Rscript 01_summarise-CARD-data.R`
Needs: `index-for-model-sequences.txt` (CARD download, in `../CARD`)
      `../beta-lactamases.txt` (list of clinically relevant beta-lactamase families)
Makes: `gene-prevalence-ncbi_plasmid.csv` etc.
    `clinical-betalactamase-groups.pdf` and `.png` figure of prevalence of given families
Currently this makes the figure only for `ncbi_wgs`. It could be changed to do it for `ncbi_chromosome` and `ncbi_plasmid` as well (trivial).

`02_make-matrix-card-data.py`
**Defunct**
_Status_: runs
_Usage_: `python 02_make-matrix-card-data.py`.
Needs: `card-genomes.txt` (CARD download, in `../CARD`)
Makes: `CARD-all-hits-presence-absence-chrom-plasmid.csv`
Could be improved to score perfect/strict matches differently. Currently both are stored as 1.

`02a02a_make-matrix-card-data.py`_make-matrix-card-data.py`
_Status_: runs
_Usage_: `python 02_make-matrix-card-data.py`.
Needs: `card-genomes.txt` (CARD download, in `../CARD`)
Makes: `CARD-all-hits-presence-absence-chrom-plasmid.csv` and other files
The version I now use.

`03a_align-families.sh`
_Status_: runs
_Usage_: `source 03a_align-families.sh` (because of conda environments, `./03a_` won't work)
Needs: `nucleotide_fasta_protein_homolog_model.fasta` (CARD download, in `../CARD`)
Extracts gene sequences from this file with `grep` (could be replaced with python to be more reliable)
Then aligns these sequences with mafft (not clustalo - )
Then uses `snp-dists -a -b` to get differences including insertions and deletions (gaps, not just SNPs)
(Could be put into python with calls, so that it can be merged with script below - but not necessary as it works now)

`03b_make-heatmap.py`
_Status_: runs
_Usage_: `python 03b_make-heatmap.py` (takes a few minutes; OXA is slow because large family)
Needs: `gene-prevalence-ncbi_wgs.csv` (or `ncbi_chromosome`, `ncbi_plasmid`). Those files are generated by `summarise-CARD-data.R`
`_diffs_aln.tsv` generated by `03a_align-families.sh`
Could save the clusters for a given distance threshold to file and/or compute the prevalence of each cluster. Might be interesting to add later.
Could omit the 'empty rows' i.e. those sequence variants that have no hits in CARD.
-> read in prevalence information first. Could filter out the empty rows pre-clustering. Or even further up, with mafft (although I don't think it should affect the result much)

## Downloading sequences

`get-gene-cluster-accessions.py`
_Status_: runs.
Needs:
`CARD-all-hits-presence-absence-chrom-plasmid.csv` (made by `make-matrix-card-data.py`, see below)
`card-genomes.txt` (CARD download, in `../CARD`)
`_diffs_aln.tsv` (from `make-heatmap.sh`)
Could be altered to take arguments with argparse.
Currently does not download the genomes.
Has getNeighbours within it.
Could call `ncbi-acc-download` from within script

## Processing

Once the fastas are all downloaded, this is what the pipeline looks like at the moment: something like

```bash
APIKEY="abf81216999e19fc8a23ae76fb9c6b208908" # for NCBI downloads - better to do this once, but tbd
FAMILY="CTX-M"
GENE="CTX-M-15"
THRESHOLD="25"
GENE_SEQUENCE="ATGGTTAAAAAATCACTGCGCCAGTTCACGCTGATGGCGACGGCAACCGTCACGCTGTTGTTAGGAAGTGTGCCGCTGTATGCGCAAACG
GCGGACGTACAGCAAAAACTTGCCGAATTAGAGCGGCAGTCGGGAGGCAGACTGGGTGTGGCATTGATTAACACAGCAGATAATTCGCAA
ATACTTTATCGTGCTGATGAGCGCTTTGCGATGTGCAGCACCAGTAAAGTGATGGCCGCGGCCGCGGTGCTGAAGAAAAGTGAAAGCGAA
CCGAATCTGTTAAATCAGCGAGTTGAGATCAAAAAATCTGACCTTGTTAACTATAATCCGATTGCGGAAAAGCACGTCAATGGGACGATG
TCACTGGCTGAGCTTAGCGCGGCCGCGCTACAGTACAGCGATAACGTGGCGATGAATAAGCTGATTGCTCACGTTGGCGGCCCGGCTAGC
GTCACCGCGTTCGCCCGACAGCTGGGAGACGAAACGTTCCGTCTCGACCGTACCGAGCCGACGTTAAACACCGCCATTCCGGGCGATCCG
CGTGATACCACTTCACCTCGGGCAATGGCGCAAACTCTGCGGAATCTGACGCTGGGTAAAGCATTGGGCGACAGCCAACGGGCGCAGCTG
GTGACATGGATGAAAGGCAATACCACCGGTGCAGCGAGCATTCAGGCTGGACTGCCTGCTTCCTGGGTTGTGGGGGATAAAACCGGCAGC
GGTGGCTATGGCACCACCAACGATATCGCGGTGATCTGGCCAAAAGATCGTGCGCCGCTGATTCTGGTCACTTACTTCACCCAGCCTCAA
CCTAAGGCAGAAAGCCGTCGCGATGTATTAGCGTCGGCGGCTAAAATCGTCACCGACGGTTTGTAA" # This would ideally be obtained from CARD programmatically - I have the data somewhere


python get-gene-cluster-accessions.py $FAMILY $GENE $THRESHOLD --singlehits

mkdir -p "$FAMILY"

# ...Download sequences from rescomp...




cd "$FAMILY"
echo ">gene\n$GENE_SEQUENCE" > gene.fa
scriptDir='../'

cat N*.fa | awk '$1!=""' > all.fa # remove empty lines
rm N*fa # remove original files

# Extract regions
echo "Extracting flanking regions..."
python "$scriptDir"/extract-region.py --gene gene.fa --input all.fa --upstream 5000 --downstream 5000 --complete --output all_u5k_d5k
# unsure if --circular is appropriate here - depends on whether plasmids and chromosomes can be trusted to be circular  
# The blast hits - some are not full-length. I don't check for this at the moment. It seems it could be caused by issues the with gap penalty etc parameters of blast. If one requests --complete, then there is an implicit check because the extracted regions must be equal to gene_length + flanking_regions

# Align gene sequences  
echo "Aligning gene sequences with mafft..."
mafft --quiet all_u5k_d5k_focal_gene.fa > all_u5k_d5k_focal_gene.aln
# Deduplicate
seqkit rmdup -s < all_u5k_d5k_focal_gene.aln > all_u5k_d5k_focal_gene.dedup.aln -D all_u5k_d5k_focal_gene.dedup.txt
# SNP distances  
echo "SNP distances..."
snp-dists -q -m all_u5k_d5k_focal_gene.aln   > pangraph_all_u5k_d5k.gfa.gene.snps.tsv   

# Pangraph
echo "Pangraph..."
pangraph build all_u5k_d5k.fa > pangraph_all_u5k_d5k.json # what should the minimum block size be?
pangraph export --edge-minimum-length 0 pangraph_all_u5k_d5k.json -p pangraph_all_u5k_d5k  -o ./
# Optional: polishing (time-intensive) and export
#pangraph polish pangraph_all_u5k_d5k.json > pangraph_all_u5k_d5k_polished.json
# pangraph export --edge-minimum-length 0 pangraph_all_u5k_d5k_polished.json -p pangraph_all_u5k_d5k_polished  -o ./
# pangraph export pangraph_all_u5k_d5k.json -p pangraph_all_u5k_d5k_polished  -o ./ --export-panX --no-export-gfa


# Prepare gfa
echo "Preparing gfa..."
python "$scriptDir"/prepare-pangraph-gfa.py pangraph_all_u5k_d5k.gfa
# change the .colours.csv output
# Should use proper lengths of blocks - but this is a minor addition for once the pipeline is done.
# this now also saves the counts of unique full-length paths and a representative of the most frequent path to
# a file

# Find the gene block in pangraph  
# (could be incorporated into the prepare pangraph step?
echo "Finding gene block..."
makeblastdb -in pangraph_all_u5k_d5k.fa -dbtype 'nucl'
geneBlock=$(blastn -query gene.fa -db pangraph_all_u5k_d5k.fa -outfmt 6 | cut -f 2)

# Use the compute-distances script
echo "Computing distances..."
python "$scriptDir"/compute-distances.py pangraph_all_u5k_d5k.gfa $geneBlock

# Plot the distances  
echo "Plotting distances..."
Rscript "$scriptDir"/plot-output-dists.R pangraph_all_u5k_d5k.gfa.output_dists.csv pangraph_all_u5k_d5k.gfa.most_frequent_path_representative.txt all_u5k_d5k_focal_gene.dedup.txt
```

`prepare-pangraph-gfa.py`
Within `pangraph-tutorials`. Although I am using a version of it in the scripts directory here.

`compute-distances.py`
_Status_: runs.
_Usage_: `python compute-distances.py pangraph_all_u5k_d5k.gfa $geneBlock`
Needs to:
Take the actual location of the gene in the alignment block into account. I think there might be some problems with overhangs in the alignment block for divergent structures (can be checked with the panX export - looking at this for the central block should be a standard diagnostic output). These should also contain the tandem dinucleotide repeats - is there a way to systematically look for these like with TEtyper but without a specified background structure?
Use the actual lengths of the blocks in the sequence? (downside: wouldn't then be symmetric between sequences)
Compute: number of block breakpoints; sequence that is in blocks not seenin the other sequence


`plot-output-dists.R`
_Status_: runs.
_Usage_: `Rscript plot-output-dists.R pangraph_all_u5k_d5k.gfa.output_dists.csv plot-output.pdf`

Could compute SNP distances within python rather than snp-dists.
Will have the gene sequences, and the extracted regions, and the pangraph block information.

### Visualization

`plot-breakpoint-dists-SNPs.R`
_Status_: not adapted yet from `CTX-M-55/true` example

## CARD data files

These are now in `beta-lactamases/CARD` since we don't need all CARD download files.

From CARD-prevalence download v3.1.0 [here](https://card.mcmaster.ca/download/6/prevalence-v3.1.0.tar.bz2):
`index-for-model-sequences.txt.gz`: contains all the detection statistics for the
sequences available in the above FASTA files, indicating pathogen, detection criteria,
ARO categorization, and similarity to curated CARD reference sequence

`card-genomes.txt.gz`: tab-separated; indicates for each analyzed genome which putative AMR
determinants were detected by RGI at both the Perfect and Strict confidence levels. This
provides a "putative resistome"  for each genome. Available and viewable on the web at:
 https://card.mcmaster.ca/genomes

From CARD-data download v3.2.3 [here](https://card.mcmaster.ca/download/0/broadstreet-v3.2.3.tar.bz2):

`nucleotide_fasta_protein_homolog_model.fasta`: Nucleotide and corresponding protein FASTA downloads are available as separate files for each model type.  For example, the "protein homolog" model type contains sequences of antimicrobial resistance genes that do not include mutation as a determinant of resistance - these data are appropriate for BLAST analysis of metagenomic data or searches excluding secondary screening for resistance mutations.
