# Details of scripts

Unless otherwise specified, all should be run from within `scripts`. Remember to activate the conda environment first with `conda activate betalactamases` (or have all necessary things installed).

## Default pipeline

The default pipeline `runDefaultPipeline.py` is a python script that calls other scripts. It will take a multi-fasta file of sequences which are presumed to all have a gene in common (e.g. an antibiotic resistance gene) present in one copy only. The pipeline will extract a region of specified size (default: 5000 bases upstream and 5000 bases downstream) and then run pangraph and produce diagnostic visualizations of the diversity around the central gene. 

```
usage: runDefaultPipeline [-h] --fasta FASTA --gene GENE [--outputdir OUTPUTDIR] [--upstream UPSTREAM] [--downstream DOWNSTREAM] [--polish] [--panx] [--bandage] [--prefix PREFIX]

Run pangraph pipeline on the flanking regions of a gene.

options:
  -h, --help            show this help message and exit
  --fasta FASTA         Multi-fasta file of input sequences containing the gene.
  --gene GENE           Fasta of focal gene to centre analysis on.
  --outputdir OUTPUTDIR
                        Output directory (default='./')
  --upstream UPSTREAM   Upstream bases (default=5000)
  --downstream DOWNSTREAM
                        Upstream bases (default=5000)
  --polish              Whether to use pangraph polish (time-intensive; default=False).
  --panx                Whether to export panX output (time-intensive; default=False).
  --bandage             Whether to run Bandage to get Bandage-visualized graph (default=False)
  --prefix PREFIX       Output prefix (default=current date & time) 
```

The plotting component scripts require the following R libraries: `RColorBrewer, cowplot, dplyr, ggdendro, gggenes, ggplot2, ggridges, ggtree, reshape2, vegan` (N.B. not checked to see if some are not in fact required). 

Assuming that all the fasta files have been downloaded, to prepare a multi-fasta file of input sequences containing a given named gene variant (and nearby genes) one can use `makeMultiFasta.py`:

```
usage: makeMultiFasta [-h] --gene GENE [--threshold THRESHOLD] --fastadir FASTADIR [--outputdir OUTPUTDIR] [--multiplehits]

Make a multi fasta file for a given gene.

options:
  -h, --help            show this help message and exit
  --gene GENE           Name of gene (e.g. CTX-M-15, VIM-1, NDM-1).
  --threshold THRESHOLD
                        SNP threshold for distance from gene variant.
  --fastadir FASTADIR   Directory with fasta files.
  --outputdir OUTPUTDIR
                        Output directory
  --multiplehits        Whether to allow multiple hits
```


## Plotting

Bandage plots can be generated with a standalone script, assuming that previous data has been generated by pipeline (can be handy if wanting to adjust them with tweaked image size etc.):

```
usage: bandagePlot [-h] --prefix PREFIX [--bandagewidth BANDAGEWIDTH]
                   [--bandageheight BANDAGEHEIGHT] [--width WIDTH] [--height HEIGHT]
                   [--output OUTPUT]

Plots bandage and block plot for given input files.

options:
  -h, --help            show this help message and exit
  --prefix PREFIX       Prefix of files to use
  --bandagewidth BANDAGEWIDTH
                        Width of Bandage image (px)
  --bandageheight BANDAGEHEIGHT
                        Height of Bandage image (px)
  --width WIDTH         Width of final plot (inches)
  --height HEIGHT       Height of final plot (inches)
  --output OUTPUT       Output prefix
```

For some supplementary plotting with metadata to remove comparisons which have identical country & year (so may be from outbreak sequencing investigations) one can use 

```

for f in $(ls -d /Users/Liam/Dropbox/_Projects/pangraph/data/*/); # N.B. path would need adjusting
do
  echo $f
  prefix=$(echo $f | rev | cut -d '/' -f 2 | rev | cut -d '-' -f 4-6)
  echo $prefix
  file1=$(ls $f*.output_dists.csv)
  file2=$(ls $f*most_frequent_path_representative.txt)
  file3=$(ls $f*focal_gene.dedup.txt)
  Rscript plot-output-dists.R $file1 $file2 $file3 
  Rscript plot-output-dists-use-metadata.R $file1 $file2 $file3
done

```

## Annotations

For looking at annotations, one can use work-in-progress (3 Feb 2023) script:

```
usage: testPlotAnnotationsAMR [-h] --prefix PREFIX --annotations ANNOTATIONS [--width WIDTH]
                              [--height HEIGHT] --output OUTPUT

Plots blocks with annotations: currently AMR annotations.

options:
  -h, --help            show this help message and exit
  --prefix PREFIX       Prefix of files to use
  --annotations ANNOTATIONS
                        File of annotations (currently assumed generated with abricate)
  --width WIDTH         Width of final plot (inches)
  --height HEIGHT       Height of final plot (inches)
  --output OUTPUT       Output pdf
```